File: Sources/iTunesKit/Services/iTunesWebTokenService.swift
import Foundation

/// A service that scrapes the Apple Music web player to obtain a developer token.
public actor iTunesWebTokenService {
    
    private let session: URLSession
    private var cachedToken: String?
    private var tokenExpiry: Date?
    private var lastScrapeDate: Date?
    
    // Minimum time between scrapes to avoid spamming Apple's servers
    private let minScrapeInterval: TimeInterval = 3600 // 1 hour
    
    public init(session: URLSession = .shared) {
        self.session = session
    }
    
    /// Fetches a valid developer token by scraping the Apple Music application script.
    public func fetchToken() async throws -> String {
        // 1. Return cached token if still valid
        if let token = cachedToken, let expiry = tokenExpiry, expiry > Date() {
            return token
        }
        
        // 2. Check if we scraped too recently
        if let lastScrape = lastScrapeDate, Date().timeIntervalSince(lastScrape) < minScrapeInterval {
            print("⚠️ iTunesKit: Rate limiting scrape. Using last known token (or failing if none).")
            if let token = cachedToken { return token }
            throw URLError(.cannotConnectToHost) // Or a custom rate limit error
        }
        
        print("🔍 iTunesKit: Scraping Apple Music Web Player...")
        
        guard let url = URL(string: Secrets.webNewURL) else {
            throw URLError(.badURL)
        }
        
        // Mark scrape attempt
        self.lastScrapeDate = Date()
        
        // Fetch the bootstrap HTML to find the latest application script
        var request = URLRequest(url: url)
        request.setValue(Secrets.userAgent, forHTTPHeaderField: "User-Agent")
        
        let (data, _) = try await session.data(for: request)
        guard let html = String(data: data, encoding: .utf8) else {
             throw URLError(.cannotDecodeRawData)
        }
        
        // Find the main application JS file (e.g., /assets/index~[hash].js)
        let jsPattern = "/assets/index~[a-zA-Z0-9]+\\.js"
        guard let jsRange = html.range(of: jsPattern, options: .regularExpression) else {
            print("❌ iTunesKit: Could not find application script URL in HTML")
            throw URLError(.cannotParseResponse)
        }
        
        let jsPath = String(html[jsRange])
        let jsURLString = "\(Secrets.webURL)\(jsPath)"
        print("🚀 iTunesKit: Fetching application script: \(jsURLString)")
        
        guard let jsURL = URL(string: jsURLString) else {
            throw URLError(.badURL)
        }
        
        var jsRequest = URLRequest(url: jsURL)
        jsRequest.setValue(Secrets.userAgent, forHTTPHeaderField: "User-Agent")
        
        let (jsData, _) = try await session.data(for: jsRequest)
        guard let jsContent = String(data: jsData, encoding: .utf8) else {
            throw URLError(.cannotDecodeRawData)
        }
        
        if let token = findToken(in: jsContent) {
            // Success! Cache the token for 2 hours (typical Apple Music token lifespan is longer, but 2h is safe)
            self.cachedToken = token
            self.tokenExpiry = Date().addingTimeInterval(7200) 
            print("✅ iTunesKit: Successfully obtained devToken from script")
            return token
        }

        print("❌ iTunesKit: Failed to find devToken in application script")
        throw URLError(.cannotParseResponse)
    }
    
    private func findToken(in text: String) -> String? {
        let pattern = "eyJ[a-zA-Z0-9._-]{200,}" 
        guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else { return nil }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        let matches = regex.matches(in: text, options: [], range: range)
        
        for match in matches {
            if let tokenRange = Range(match.range, in: text) {
                let token = String(text[tokenRange])
                let dotCount = token.filter { $0 == "." }.count
                if dotCount >= 2 {
                    return token
                }
            }
        }
        return nil
    }
}

File: ituneskit/src/main/java/aaravgupta/ituneskit/services/iTunesWebTokenService.kt
package aaravgupta.ituneskit

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

public class iTunesWebTokenService(
    private val session: OkHttpClient = NetworkDefaults.defaultOkHttpClient(),
    cacheFile: File? = null,
    private val cacheDirectoryProvider: iTunesCacheDirectoryProvider = SystemCacheDirectoryProvider,
    private val nowProvider: () -> Long = { System.currentTimeMillis() },
    private val logger: (String) -> Unit = { println(it) }
) {
    @Serializable
    private data class CachedTokenState(
        val token: String? = null,
        val tokenExpiry: String? = null,
        val lastScrapeDate: String? = null
    )

    private val json: Json = NetworkDefaults.defaultJson
    private val mutex: Mutex = Mutex()
    private val cacheFile: File = cacheFile ?: defaultCacheFile(cacheDirectoryProvider)

    private var cachedToken: String? = null
    private var tokenExpiryMillis: Long? = null
    private var lastScrapeDateMillis: Long? = null

    init {
        val state = loadState()
        if (state != null) {
            cachedToken = state.token
            tokenExpiryMillis = state.tokenExpiry?.let(Iso8601DateCodec::parse)
            lastScrapeDateMillis = state.lastScrapeDate?.let(Iso8601DateCodec::parse)
        }
    }

    public suspend fun fetchToken(): String = mutex.withLock {
        val now = nowProvider()

        validCachedToken(now)?.let { return@withLock it }

        val lastScrapeDate = lastScrapeDateMillis
        if (lastScrapeDate != null && (now - lastScrapeDate) < MIN_SCRAPE_INTERVAL_MILLIS) {
            logger("iTunesKit: Rate limiting scrape. Using last known token (or failing if none).")
            validCachedToken(now)?.let { return@withLock it }
            throw iTunesKitException.CannotConnectToHost(
                "Scrape throttled and no valid cached token available"
            )
        }

        logger("iTunesKit: Scraping Apple Music Web Player...")
        if (Secrets.webNewURL.toHttpUrlOrNull() == null) {
            throw iTunesKitException.BadUrl(Secrets.webNewURL)
        }

        val scrapeDate = now
        lastScrapeDateMillis = scrapeDate
        persistState()

        val html = fetchText(
            url = Secrets.webNewURL,
            headers = mapOf("User-Agent" to Secrets.userAgent)
        )

        val jsPath = extractJavaScriptPath(html)
            ?: throw iTunesKitException.CannotParseResponse()

        val jsURL = "${Secrets.webURL}$jsPath"
    logger("iTunesKit: Fetching application script: $jsURL")

        val jsContent = fetchText(
            url = jsURL,
            headers = mapOf("User-Agent" to Secrets.userAgent)
        )

        val token = findToken(jsContent)
            ?: throw iTunesKitException.CannotParseResponse()

        cachedToken = token
        tokenExpiryMillis = nowProvider() + TOKEN_LIFETIME_MILLIS
        lastScrapeDateMillis = scrapeDate
        persistState()

        logger("iTunesKit: Successfully obtained devToken from script")
        token
    }

    private suspend fun fetchText(url: String, headers: Map<String, String>): String {
        val parsedURL = url.toHttpUrlOrNull() ?: throw iTunesKitException.BadUrl(url)
        val requestBuilder = Request.Builder().url(parsedURL).get()

        for ((name, value) in headers) {
            requestBuilder.header(name, value)
        }

        val request = requestBuilder.build()

        return withContext(Dispatchers.IO) {
            session.newCall(request).execute().use { response ->
                val responseBytes = response.body?.bytes() ?: ByteArray(0)
                if (!response.isSuccessful) {
                    throw iTunesKitException.BadServerResponse(
                        statusCode = response.code,
                        url = url,
                        responseBody = responseBytes.toString(Charsets.UTF_8).ifBlank { null }
                    )
                }

                responseBytes.toString(Charsets.UTF_8)
            }
        }
    }

    private fun extractJavaScriptPath(html: String): String? {
        return JS_BUNDLE_PATTERN.find(html)?.value
    }

    private fun findToken(text: String): String? {
        for (match in TOKEN_PATTERN.findAll(text)) {
            val token = match.value
            if (token.count { it == '.' } >= 2) {
                return token
            }
        }
        return null
    }

    private fun validCachedToken(now: Long): String? {
        val token = cachedToken
        val expiry = tokenExpiryMillis
        if (token != null && expiry != null && expiry > now) {
            return token
        }
        return null
    }

    private fun persistState() {
        val state = CachedTokenState(
            token = cachedToken,
            tokenExpiry = tokenExpiryMillis?.let(Iso8601DateCodec::format),
            lastScrapeDate = lastScrapeDateMillis?.let(Iso8601DateCodec::format)
        )

        try {
            cacheFile.parentFile?.mkdirs()
            val encoded = json.encodeToString(CachedTokenState.serializer(), state)
            cacheFile.writeText(encoded)
        } catch (error: Throwable) {
            logger("iTunesKit: Failed to persist devToken cache: $error")
        }
    }

    private fun loadState(): CachedTokenState? {
        return try {
            if (!cacheFile.exists()) {
                null
            } else {
                val text = cacheFile.readText()
                json.decodeFromString(CachedTokenState.serializer(), text)
            }
        } catch (_: Throwable) {
            null
        }
    }

    public companion object {
        private const val MIN_SCRAPE_INTERVAL_MILLIS: Long = 3_600_000L
        private const val TOKEN_LIFETIME_MILLIS: Long = 7_200_000L

        private val JS_BUNDLE_PATTERN = Regex("/assets/index~[a-zA-Z0-9]+\\.js")
        private val TOKEN_PATTERN = Regex("eyJ[a-zA-Z0-9._-]{200,}")

        @Volatile
        private var sharedCacheDirectoryProvider: iTunesCacheDirectoryProvider = SystemCacheDirectoryProvider

        @Volatile
        private var sharedInstance: iTunesWebTokenService? = null

        private fun defaultCacheFile(provider: iTunesCacheDirectoryProvider): File {
            return provider.cacheDirectory()
                .resolve("iTunesKit")
                .resolve("web-token-cache.json")
        }

        @JvmStatic
        public fun configureSharedCacheDirectoryProvider(provider: iTunesCacheDirectoryProvider) {
            synchronized(this) {
                sharedCacheDirectoryProvider = provider
                sharedInstance = null
            }
        }

        @JvmStatic
        public val shared: iTunesWebTokenService
            get() = synchronized(this) {
                sharedInstance ?: iTunesWebTokenService(
                    cacheDirectoryProvider = sharedCacheDirectoryProvider
                ).also { sharedInstance = it }
            }
    }
}

private object Iso8601DateCodec {
    private val utcTimeZone: TimeZone = TimeZone.getTimeZone("UTC")
    private val formats: List<String> = listOf(
        "yyyy-MM-dd'T'HH:mm:ss.SSSX",
        "yyyy-MM-dd'T'HH:mm:ssX"
    )

    fun format(epochMillis: Long): String {
        val formatter = SimpleDateFormat(formats.first(), Locale.US)
        formatter.timeZone = utcTimeZone
        return formatter.format(Date(epochMillis))
    }

    fun parse(value: String): Long? {
        for (pattern in formats) {
            val formatter = SimpleDateFormat(pattern, Locale.US)
            formatter.timeZone = utcTimeZone
            val parsed = formatter.parse(value)
            if (parsed != null) {
                return parsed.time
            }
        }

        return null
    }
}